File: frontend/src/types/migration.ts
import { t } from "../theme";

export type Strategy = "CDC_STAGE" | "CDC_DIRECT" | "BULK_STAGE" | "BULK_DIRECT";

export const hasCdc    = (s: Strategy): boolean => s.startsWith("CDC_");
export const usesStage = (s: Strategy): boolean => s.endsWith("_STAGE");

export const strategyLabel = (s: Strategy): string =>
  `${hasCdc(s) ? "С CDC" : "Без CDC"} (${usesStage(s) ? "stage" : "direct"})`;

export const composeStrategy = (cdc: boolean, stage: boolean): Strategy =>
  `${cdc ? "CDC" : "BULK"}_${stage ? "STAGE" : "DIRECT"}` as Strategy;

export type MigrationPhase =
  | "DRAFT" | "NEW"
  | "TOPIC_CREATING"
  | "CHUNKING" | "BULK_LOADING" | "BULK_LOADED"
  | "STAGE_VALIDATING" | "STAGE_VALIDATED"
  | "BASELINE_PUBLISHING" | "BASELINE_LOADING" | "BASELINE_PUBLISHED"
  | "STAGE_DROPPING" | "INDEXES_ENABLING"
  | "DATA_VERIFYING" | "DATA_MISMATCH"
  | "CDC_APPLY_STARTING" | "CDC_APPLYING" | "CDC_CATCHING_UP" | "CDC_CAUGHT_UP"
  | "STEADY_STATE" | "PAUSED"
  | "CANCELLING" | "CANCELLED"
  | "COMPLETED" | "FAILED";

export interface Migration {
  migration_id: string;
  migration_name: string;
  phase: MigrationPhase;
  state_changed_at: string;
  source_connection_id: string;
  target_connection_id: string;
  source_schema: string;
  source_table: string;
  target_schema: string;
  target_table: string;
  stage_table_name: string;
  stage_tablespace: string;
  connector_name: string;
  topic_prefix: string;
  consumer_group: string;
  chunk_strategy: string;
  chunk_size: number;
  apply_mode: string;
  source_pk_exists: boolean;
  source_uk_exists: boolean;
  effective_key_type: string;
  effective_key_source: string;
  effective_key_columns_json: string;
  key_uniqueness_validated: boolean;
  key_validation_status: string | null;
  key_validation_message: string | null;
  start_scn: string | null;
  scn_fixed_at: string | null;
  created_by: string | null;
  description: string | null;
  created_at: string;
  updated_at: string;
  locked_by: string | null;
  lock_until: string | null;
  error_code: string | null;
  error_text: string | null;
  failed_phase: string | null;
  retry_count: number;
  // ── Progress / monitoring (added stage 1) ────────────────────────────────
  total_rows: number | null;
  total_chunks: number | null;
  chunks_done: number;
  chunks_failed: number;
  validate_hash_sample: boolean;
  validation_result: Record<string, unknown> | null;
  connector_status: string | null;
  kafka_lag: number | null;
  kafka_lag_checked_at: string | null;
  rows_loaded: number;
  max_parallel_workers: number;
  baseline_parallel_degree: number;
  baseline_batch_size: number;
  strategy: Strategy;
  baseline_chunks_total: number | null;
  baseline_chunks_done: number;
  queue_position: number | null;
  group_id: string | null;
  data_compare_task_id?: string;
  truncate_target: boolean;
}

// ── Connector Groups ──────────────────────────────────────────────────────────

export type GroupStatus = "PENDING" | "TOPICS_CREATING" | "CONNECTOR_STARTING" | "RUNNING" | "STOPPING" | "STOPPED" | "FAILED";

export interface ConnectorGroup {
  group_id: string;
  group_name: string;
  source_connection_id: string;
  connector_name: string;
  topic_prefix: string;
  consumer_group_prefix: string;
  status: GroupStatus;
  error_text: string | null;
  table_include_list?: string;
  message_key_columns?: string;
  created_at: string;
  updated_at: string;
  migrations?: MigrationSummary[];
}

export interface MigrationSummary {
  migration_id: string;
  migration_name: string;
  phase: MigrationPhase;
  state_changed_at: string;
  source_schema: string;
  source_table: string;
  target_schema: string;
  target_table: string;
  created_at: string;
  updated_at: string;
  error_code: string | null;
  error_text: string | null;
  failed_phase: string | null;
  retry_count: number;
  description: string | null;
  created_by: string | null;
  total_rows: number | null;
  total_chunks: number | null;
  chunks_done: number;
  chunks_failed: number;
  rows_loaded: number;
  strategy: Strategy;
  queue_position: number | null;
}

export interface StateHistoryEntry {
  id: number;
  migration_id: string;
  from_phase: string | null;
  to_phase: string;
  transition_status: string;
  transition_reason: string | null;
  message: string | null;
  actor_type: string;
  actor_id: string | null;
  correlation_id: string | null;
  created_at: string;
}

export interface MigrationDetail extends Migration {
  history: StateHistoryEntry[];
}

// ── Phase colours ────────────────────────────────────────────────────────────

interface PhaseColor { bg: string; text: string; border: string }

const PHASE_COLORS: Record<string, PhaseColor> = {
  DRAFT:               { bg: t.border.subtle, text: t.text.secondary, border: t.border.base },
  NEW:                 { bg: t.bg.s3, text: t.blue.fg, border: t.blue.dim },
  TOPIC_CREATING:      { bg: t.purple.bg, text: t.purple.fg, border: t.purple.base },
  CHUNKING:            { bg: t.amber.bg, text: t.amber.fg, border: t.amber.dim },
  BULK_LOADING:        { bg: t.amber.bg, text: t.amber.fg, border: t.amber.dim },
  BULK_LOADED:         { bg: t.amber.bg, text: t.amber.fg, border: t.amber.dim },
  STAGE_VALIDATING:    { bg: t.blue.bg, text: t.blue.fg, border: t.blue.dim },
  STAGE_VALIDATED:     { bg: t.blue.bg, text: t.blue.fg, border: t.blue.dim },
  BASELINE_PUBLISHING: { bg: t.purple.bg, text: t.purple.fg, border: t.purple.base },
  BASELINE_LOADING:    { bg: t.purple.bg, text: t.purple.fg, border: t.purple.base },
  BASELINE_PUBLISHED:  { bg: t.purple.bg, text: t.purple.fg, border: t.purple.base },
  STAGE_DROPPING:      { bg: t.bg.s2, text: t.green.fg, border: t.green.base },
  INDEXES_ENABLING:    { bg: t.bg.s2, text: t.green.fg, border: t.green.base },
  DATA_VERIFYING:      { bg: t.blue.bg, text: t.blue.fg, border: t.blue.dim },
  DATA_MISMATCH:       { bg: t.red.bg, text: t.amber.fg, border: t.amber.dim },
  CDC_APPLY_STARTING:  { bg: t.red.bg, text: t.amber.fg, border: t.amber.dim },
  CDC_APPLYING:        { bg: t.red.bg, text: t.amber.fg, border: t.amber.dim },
  CDC_CATCHING_UP:     { bg: t.red.bg, text: t.amber.fg, border: t.amber.dim },
  CDC_CAUGHT_UP:       { bg: t.red.bg, text: t.amber.fg, border: t.amber.dim },
  STEADY_STATE:        { bg: t.green.bg, text: t.green.fg, border: t.green.dim },
  PAUSED:              { bg: t.border.subtle, text: t.text.primary, border: t.text.disabled },
  CANCELLING:          { bg: t.red.bg, text: t.red.fg, border: t.red.dim },
  CANCELLED:           { bg: t.bg.s2, text: t.text.muted, border: t.text.disabled },
  COMPLETED:           { bg: t.green.bg, text: t.green.fg, border: t.green.dim },
  FAILED:              { bg: t.red.bg, text: t.red.fg, border: t.red.dim },
};

const FALLBACK_COLOR: PhaseColor = { bg: t.border.subtle, text: t.text.secondary, border: t.border.base };

export function phaseColor(phase: string): PhaseColor {
  return PHASE_COLORS[phase] ?? FALLBACK_COLOR;
}

export const ORDERED_PHASES: MigrationPhase[] = [
  "DRAFT", "NEW",
  "TOPIC_CREATING",
  "CHUNKING", "BULK_LOADING", "BULK_LOADED",
  "STAGE_VALIDATING", "STAGE_VALIDATED",
  "BASELINE_PUBLISHING", "BASELINE_LOADING", "BASELINE_PUBLISHED",
  "STAGE_DROPPING", "INDEXES_ENABLING",
  "DATA_VERIFYING", "DATA_MISMATCH",
  "CDC_APPLY_STARTING", "CDC_APPLYING", "CDC_CATCHING_UP", "CDC_CAUGHT_UP",
  "STEADY_STATE",
];

File: backend/models/migration.py
from __future__ import annotations

import json
from datetime import datetime
from decimal import Decimal
from typing import Any

from pydantic import BaseModel, ConfigDict, field_validator

from models.enums import (
    ChunkStatus,
    ChunkType,
    MigrationMode,
    MigrationStrategy,
    Phase,
)


class Migration(BaseModel):
    model_config = ConfigDict(from_attributes=True, populate_by_name=True)

    migration_id: str
    migration_name: str
    phase: Phase = Phase.DRAFT
    state_changed_at: datetime | None = None

    source_connection_id: str = ""
    target_connection_id: str = ""
    source_schema: str = ""
    source_table: str = ""
    target_schema: str = ""
    target_table: str = ""

    stage_table_name: str = ""
    stage_tablespace: str = ""
    connector_name: str = ""
    topic_prefix: str = ""
    consumer_group: str = ""

    chunk_strategy: str = ""
    chunk_size: int = 10000
    max_parallel_workers: int = 1
    apply_mode: str = ""

    source_pk_exists: bool = False
    source_uk_exists: bool = False
    effective_key_type: str = ""
    effective_key_source: str = ""
    effective_key_columns_json: str = "[]"

    key_uniqueness_validated: bool = False
    key_validation_status: str | None = None
    key_validation_message: str | None = None

    start_scn: Decimal | None = None
    scn_fixed_at: datetime | None = None

    created_by: str | None = None
    description: str | None = None
    created_at: datetime | None = None
    updated_at: datetime | None = None

    locked_by: str | None = None
    lock_until: datetime | None = None

    error_code: str | None = None
    error_text: str | None = None
    failed_phase: str | None = None
    retry_count: int = 0

    total_rows: int | None = None
    total_chunks: int | None = None
    chunks_done: int = 0
    chunks_failed: int = 0
    validate_hash_sample: bool = False
    validation_result: dict[str, Any] | None = None
    connector_status: str | None = None
    kafka_lag: int | None = None
    kafka_lag_checked_at: datetime | None = None
    rows_loaded: int = 0

    baseline_parallel_degree: int = 1
    baseline_batch_size: int = 500000
    migration_strategy: MigrationStrategy = MigrationStrategy.STAGE
    baseline_chunks_total: int | None = None
    baseline_chunks_done: int = 0
    queue_position: int | None = None
    migration_mode: MigrationMode = MigrationMode.CDC
    data_compare_task_id: str | None = None
    group_id: str | None = None

    @field_validator("validation_result", mode="before")
    @classmethod
    def _parse_validation_result(cls, v: Any) -> dict[str, Any] | None:
        if isinstance(v, str):
            return json.loads(v)
        return v

    @property
    def effective_key_columns(self) -> list[str]:
        return json.loads(self.effective_key_columns_json)

    @property
    def is_terminal(self) -> bool:
        return self.phase in Phase.terminal()


class MigrationChunk(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    chunk_id: str
    migration_id: str
    chunk_seq: int
    rowid_start: str
    rowid_end: str
    status: ChunkStatus = ChunkStatus.PENDING
    rows_loaded: int = 0
    worker_id: str | None = None
    claimed_at: datetime | None = None
    started_at: datetime | None = None
    completed_at: datetime | None = None
    error_text: str | None = None
    retry_count: int = 0
    created_at: datetime | None = None
    chunk_type: ChunkType = ChunkType.BULK


class CdcState(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    migration_id: str
    consumer_group: str = ""
    topic: str = ""
    total_lag: int = 0
    lag_by_partition: dict[str, Any] | None = None
    last_event_scn: Decimal | None = None
    last_event_ts: datetime | None = None
    apply_rate_rps: Decimal | None = None
    rows_applied: int = 0
    worker_id: str | None = None
    worker_heartbeat: datetime | None = None
    updated_at: datetime | None = None


class StateHistoryEntry(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    migration_id: str
    from_phase: str | None = None
    to_phase: str
    transition_status: str = "SUCCESS"
    transition_reason: str | None = None
    message: str | None = None
    actor_type: str = "SYSTEM"
    actor_id: str | None = None
    correlation_id: str | None = None
    created_at: datetime | None = None
